### PR TITLE
Add generic device server implementation and refactor progress notifications

### LINKED ISSUE #754: Unify VirtualInterface and GatewayAgent
There is no longer a need to have a distinct VirtualInterface and GatewayAgent since both now have access to an event loop.

### LINKED ISSUE #756: Document event object format for all events raised by AbstractDeviceAdapter
We need to properly document what kind of object must be sent in each type of event and what attributes or keys it must contain for interoperability.  

There are a lot of implicit rules about scan payloads for example that we should collect and document in terms of how the expiration time needs to be reported.

### PR DESCRIPTION
**Pull request by timburke**
## Overview

Adds  `AbstractDeviceServer` and `StandardDeviceServer` classes.  These classes are designed to unify and eventually replace `VirtualInterface` and `GatewayAgent`.  

`AbstractDeviceServer` is designed to let `iotile-gateway` load and run device servers without needing to know anything about how they serve devices.  It is intentionally left very minimal.

`StandardDeviceServer` is designed to be the reference implementation of device server functionality that most/all device servers inherit from.  As described in the documentation, it takes a single `AbstractDeviceAdapter` instance and serves multiple clients access to the devices that are reachable through that adapter.

`StandardDeviceServer` takes care of all of the tricky house-keeping around cleaning up after clients that unexpectedly go away and leave device connections hanging.  The goal is to have all that is required to implement a new DeviceServer be just listening for connections (over http, tcp, etc) and translating the messages received into calls on `StandardDeviceServer`.

## Major Changes

- Add `AbstractDeviceServer` class
- Add `StandardDeviceServer` class that implements `AbstractDeviceServer`
- Refactor how progress notifications are sent to not be callback based but to use the same notification system as other events.

### Additional Notes

- Slightly reorganizes the `transport` subpackage to be more logical.  
- Updates `SynchronousLegacyWrapper` to explicitly inherit from DeviceAdapter and provide both `_sync` and `_async` versions of all legacy functions.  It takes care of translating the new `progress` events back to the old progress callbacks.
- Slight refactor on `BackgroundEventLoop.launch_coroutine` to allow it to be called from within and without of the loop itself.  It now returns an `asyncio.Future` or `concurrent.futures.Future` based on the calling thread.  This lets us simplify the implementation of `notify_events`.
- Documents all of the events that can be notified by an AbstractDeviceAdapter.

Closes #756
Closes #754

### CODE REVIEW

File: iotilecore/iotile/core/hw/transport/adapter/abstract.py [hunk 1]
```diff
@@ -50,14 +133,15 @@ def can_connect(self):
...
-        because a user can just try to connect to a device and have it fail if
-        the adapter cannot accomodate another connection.  However, in more
-        complicated scenarios where there are multiple DeviceAdapters that
-        could be used to connect to a device, it is useful to be able to
-        filter them and find the best adapter that has a free connection
-        slot available.
+        given DeviceAdapter may be able to maintain simultaneous connections
+        to more than one device at a time.  In simple cases, this method is
+        not necessary because a user can just try to connect to a device and
+        have it fail if the adapter cannot accomodate another connection.
```
Comment (mattrunchey) [inline iotilecore/iotile/core/hw/transport/adapter/abstract.py]:
```suggestion
        have it fail if the adapter cannot accommodate another connection.
```

File: iotilecore/iotile/core/hw/transport/adapter/abstract.py [hunk 2]
```diff
@@ -419,18 +506,19 @@ def remove_monitor(self, handle):
             RPCErrorCode: The RPC implementation wishes to fail with a
                 non-zero status code.
             RPCInvalidIDError: The rpc_id is too large to fit in 16-bits.
+            TileBusSerror: The tile was busy and could not respond to the RPC.
```
Comment (mattrunchey) [inline iotilecore/iotile/core/hw/transport/adapter/abstract.py]: Is there a typo here? That "S" sticks out

File: iotilecore/iotile/core/hw/transport/adapter/mixin_notifications.py [hunk 3]
```diff
@@ -190,8 +196,35 @@ def remove_monitor(self, handle):
...
+        """Notify an event.
+
+        This will move the notification to the background event loop and
+        return immediately.  It is useful for situations where you cannot
+        await notify_event but keep in mind that it prevents back-pressure
+        when you are notifying too fast so should be used sparingly.
+
+        Note that calling this method will push the notification to a
+        background task so it can be difficult to reason about when it will
+        precisely occur.  For that reason, :meth:`notify_event` should be
```
Comment (mattrunchey) [inline iotilecore/iotile/core/hw/transport/adapter/mixin_notifications.py]: The language here suggests that we should prefer an alternate method, but it names the same method.

File: iotilecore/iotile/core/hw/transport/adapter/abstract.py (hunk 2 above)
Comment (timburke) [reply iotilecore/iotile/core/hw/transport/adapter/abstract.py]: yeah, `TileBusyError`

File: iotilecore/iotile/core/hw/transport/adapter/mixin_notifications.py (hunk 3 above)
Comment (timburke) [reply iotilecore/iotile/core/hw/transport/adapter/mixin_notifications.py]: Good catch, updating the docstring.

**Pull request merged**